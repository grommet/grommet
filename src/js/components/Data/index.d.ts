import * as React from 'react';
import { BoxProps } from '../Box/index';

export interface View {
  name?: string;
  page?: number; // when paging
  properties?:
    | string[]
    | {
        [property: string]:
          | undefined
          | boolean
          | (string | number | boolean)[]
          | { max: number; min: number };
      };
  search?: string;
  sort?: {
    property: string;
    direction: 'asc' | 'desc';
  };
  step?: number; // page size
  columns?: string[];
}

export interface DataProps {
  id?: string;

  data: object[];
  total?: number;
  // when paging
  filteredTotal?: number;

  defaultView?: View;
  view?: string | View;
  onView?: (view: View) => void;

  // whether to render a Toolbar
  toolbar?: boolean | 'search' | 'filters' | 'view';

  properties?:
    | string[]
    | {
        [key: string]: {
          badge?: boolean;
          filter?: boolean;
          // for DataTable column header, DataFilter label, DataTableColumns label
          label?: string | React.ReactNode;
          // DataFilter options
          options?: (
            | string
            | number
            | {
                label: string;
                value: string | number | boolean;
              }
          )[];
          // DataFilter range
          range?: {
            max?: number;
            min?: number;
            step?: number;
          };
          // for internal filtering only, should searching evaluate this property
          search?: boolean;
          sort?: boolean;
        };
      };

  messages?: {
    dataFilters?: {
      clear?: string;
      heading?: string;
      open?: string;
      openSet?: {
        singular?: string;
        plural?: string;
      };
    };
    dataForm?: {
      submit?: string;
    };
    dataSearch?: {
      label?: string;
      open?: string;
    };
    dataSort?: {
      ascending?: string;
      by?: string;
      descending?: string;
      direction?: string;
      open?: string;
    };
    dataSummary?: {
      filtered?: string;
      filteredSingle?: string;
      items?: string;
      itemsSingle?: string;
      selected?: string;
      total?: string;
      totalSingle?: string;
    };
    dataTableColumns?: {
      open?: string;
      order?: string;
      select?: string;
      tip?: string;
    };
    dataTableGroupBy?: {
      clear?: string;
      label?: string;
    };
    dataView?: {
      label?: string;
    };
  };

  views?: View[];

  // Future selection
  // select: string[];
  // onSelect: (select: string[]) => void;

  // Future manage named views
  // onCreateView?: (view: View) =>  void;
  // onDeleteView?: (string) => void;
}

type divProps = Omit<
  React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'onClick'
>;

export interface DataExtendedProps extends BoxProps, DataProps, divProps {}

declare const Data: React.FC<DataExtendedProps>;

export { Data };
