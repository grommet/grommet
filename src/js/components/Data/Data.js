import React, { useContext, useEffect, useMemo, useState } from 'react';
import { AnnounceContext } from '../../contexts';
import { Box } from '../Box';
import { DataFilters } from '../DataFilters';
import { DataSearch } from '../DataSearch';
import { DataSummary } from '../DataSummary';
import { DataView } from '../DataView';
import { Toolbar } from '../Toolbar';
import { DataContext } from '../../contexts/DataContext';
import { DataPropTypes } from './propTypes';
import { MessageContext } from '../../contexts/MessageContext';
import { filter } from './filter';

const defaultDefaultView = { search: '' };

const normalizeView = (viewProp, views) =>
  (typeof viewProp === 'string' && views?.find((v) => v.name === viewProp)) ||
  (typeof viewProp === 'object' && viewProp);

export const Data = ({
  children,
  data: dataProp = [],
  defaultView = defaultDefaultView,
  filteredTotal,
  id = 'data',
  messages,
  onView,
  properties,
  toolbar,
  total,
  view: viewProp,
  views,
  ...rest
}) => {
  const [view, setView] = useState(
    normalizeView(viewProp || defaultView, views),
  );
  useEffect(
    () => setView(normalizeView(viewProp || defaultView, views)),
    [defaultView, viewProp, views],
  );
  const [toolbarKeys, setToolbarKeys] = useState([]);

  const result = useMemo(() => {
    if (onView)
      // caller is filtering
      return {
        data: dataProp,
        total,
        filteredTotal: filteredTotal ?? dataProp?.length ?? 0,
      };
    return filter(dataProp, view, properties);
  }, [dataProp, filteredTotal, onView, properties, total, view]);

  // used by DataFilters to determine if badge should appear on Filter button
  const [filtersCleared, setFiltersCleared] = useState(true);

  const [selected, setSelected] = useState([]);

  const announce = useContext(AnnounceContext);
  const { format } = useContext(MessageContext);
  // Announce to screen readers when search or filters are
  // applied and affect the underlying result set
  useEffect(() => {
    let messageId;
    if (result.total !== result.filteredTotal) {
      if (result.filteredTotal === 1) messageId = 'dataSummary.filteredSingle';
      else messageId = 'dataSummary.filtered';
    } else if (result.total === 1) messageId = 'dataSummary.totalSingle';
    else messageId = 'dataSummary.total';

    // helps account for cases like 0 results of 1 item
    const items = format({
      id: result.total === 1 ? 'dataSummary.itemsSingle' : 'dataSummary.items',
      messages: messages?.dataSummary,
    });

    announce(
      `${format({
        id: messageId,
        messages: messages?.dataSummary,
        values: {
          filteredTotal: result.filteredTotal,
          total: result.total,
          items,
        },
      })}${
        selected > 0
          ? `, ${format({
              id: 'dataSummary.selected',
              messages: messages?.dataSummary,
              values: {
                selected,
              },
            })}`
          : ''
      }`,
    );
  }, [
    announce,
    format,
    messages?.dataSummary,
    result.filteredTotal,
    result.total,
    selected,
  ]);

  // what we use for DataContext value
  const contextValue = useMemo(() => {
    const value = {
      id,
      messages,
      properties,
      filtersCleared,
      setFiltersCleared,
      selected,
      setSelected,
      view,
      views,
      ...result,
    };

    value.clearFilters = () => {
      const nextView = { ...view };
      delete nextView.properties;
      delete nextView.page;
      // by clearing the properties from a view, it is no
      // longer reflecting the view
      delete nextView.name;
      delete nextView.view;
      setFiltersCleared(true);
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.onView = (nextView) => {
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.addToolbarKey = (key) => {
      setToolbarKeys((prevKeys) => {
        if (prevKeys.includes(key)) return prevKeys;
        return [...prevKeys, key];
      });
    };
    value.toolbarKeys = toolbarKeys;

    return value;
  }, [
    id,
    messages,
    filtersCleared,
    onView,
    properties,
    result,
    selected,
    toolbarKeys,
    view,
    views,
  ]);

  let toolbarContent;
  if (toolbar) {
    toolbarContent = [
      // TO DO theme object
      <Toolbar key="toolbar" gap="medium">
        <Toolbar>
          {(toolbar === true || toolbar === 'search') && <DataSearch />}
          {(toolbar === true || toolbar === 'filters') && <DataFilters layer />}
        </Toolbar>
        {(toolbar === true || toolbar === 'view') && <DataView />}
      </Toolbar>,
      <DataSummary key="summary" />,
    ];
  }

  return (
    <DataContext.Provider value={contextValue}>
      <Box id={id} flex={false} {...rest}>
        {toolbarContent}
        {children}
      </Box>
    </DataContext.Provider>
  );
};

Data.propTypes = DataPropTypes;
