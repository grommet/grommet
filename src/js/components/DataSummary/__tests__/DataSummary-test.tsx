import React, { useState } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { Data } from '../../Data';
import { DataFilters } from '../../DataFilters';
import { DataTable } from '../../DataTable';
import { Grommet } from '../../Grommet';
import { DataSummary } from '..';

// asserts that DataSummary has text. Previously this was set to 2 to include AnnounceContext
// but that is no longer the case with improvments to AnnounceContext in "polite" mode.
const expectDataSummary = (message: string) =>
  expect(screen.getAllByText(message)).toHaveLength(1);

const data = [{ name: 'a' }, { name: 'b' }];

const Selected = ({ messages }: { messages?: object }) => {
  const [selected, setSelected] = useState<(string | number)[]>(['a']);

  return (
    <Data data={data}>
      <DataSummary messages={messages} />
      <DataTable
        select={selected}
        onSelect={(nextSelected) => setSelected(nextSelected)}
      />
    </Data>
  );
};

describe('DataSummary', () => {
  test('renders', () => {
    const { container } = render(
      <Grommet>
        <Data data={data}>
          <DataFilters>
            <DataSummary />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders total message correctly when only 1 item', () => {
    const { asFragment } = render(
      <Grommet>
        <Data data={[{ name: 'a' }]}>
          <DataFilters>
            <DataSummary />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expectDataSummary('1 item');
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders filtered message correctly when 0 results but only 1 item', () => {
    const { asFragment } = render(
      <Grommet>
        <Data
          data={[{ name: 'a' }]}
          defaultView={{
            properties: {
              name: ['b'],
            },
          }}
        >
          <DataFilters>
            <DataSummary />
          </DataFilters>
        </Data>
      </Grommet>,
    );

    expectDataSummary('0 results of 1 item');
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render DataTable selections message in alignment with selections', () => {
    render(
      <Grommet>
        <Selected />
      </Grommet>,
    );

    // default selection
    expect(screen.getByText('1 selected')).toBeTruthy();
    const rowCheckbox = screen.getByRole('checkbox', { name: 'select b' });
    fireEvent.click(rowCheckbox);
    expect(screen.getByText('2 selected')).toBeTruthy();
  });

  test('should render messages prop', () => {
    render(
      <Grommet>
        <Selected
          messages={{
            selected: '{selected} SELECTED!',
            total: '{total} items total',
          }}
        />
      </Grommet>,
    );

    expect(screen.getByText('1 SELECTED!')).toBeTruthy();
    expect(screen.getByText('2 items total')).toBeTruthy();
  });
});
