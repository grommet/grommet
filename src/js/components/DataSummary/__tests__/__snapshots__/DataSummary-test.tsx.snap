// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`DataSummary renders 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c5 {
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c7 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c7:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c7:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus >circle,
.c7:focus >ellipse,
.c7:focus >line,
.c7:focus >path,
.c7:focus >polygon,
.c7:focus >polyline,
.c7:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus ::-moz-focus-inner {
  border: 0;
}

.c7:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) >circle,
.c7:focus:not(:focus-visible) >ellipse,
.c7:focus:not(:focus-visible) >line,
.c7:focus:not(:focus-visible) >path,
.c7:focus:not(:focus-visible) >polygon,
.c7:focus:not(:focus-visible) >polyline,
.c7:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c6 {
    margin-top: 12px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    id="data"
  >
    <form
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c1"
          >
            <span
              class="c5"
            >
              2 items
            </span>
          </div>
        </div>
        <footer
          class="c6"
        >
          <button
            class="c7"
            type="submit"
          >
            Apply filters
          </button>
        </footer>
      </div>
    </form>
  </div>
</div>
`;

exports[`DataSummary renders filtered message correctly when 0 results but only 1 item 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c5 {
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c7 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c7:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c7:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus >circle,
.c7:focus >ellipse,
.c7:focus >line,
.c7:focus >path,
.c7:focus >polygon,
.c7:focus >polyline,
.c7:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus ::-moz-focus-inner {
  border: 0;
}

.c7:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) >circle,
.c7:focus:not(:focus-visible) >ellipse,
.c7:focus:not(:focus-visible) >line,
.c7:focus:not(:focus-visible) >path,
.c7:focus:not(:focus-visible) >polygon,
.c7:focus:not(:focus-visible) >polyline,
.c7:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c6 {
    margin-top: 12px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      id="data"
    >
      <form
        class="c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <div
              class="c1"
            >
              <span
                class="c5"
              >
                0 results of 1 item
              </span>
            </div>
          </div>
          <footer
            class="c6"
          >
            <button
              class="c7"
              type="submit"
            >
              Apply filters
            </button>
          </footer>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DataSummary renders total message correctly when only 1 item 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c6 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c5 {
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c7 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c7:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c7:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus >circle,
.c7:focus >ellipse,
.c7:focus >line,
.c7:focus >path,
.c7:focus >polygon,
.c7:focus >polyline,
.c7:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus ::-moz-focus-inner {
  border: 0;
}

.c7:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) >circle,
.c7:focus:not(:focus-visible) >ellipse,
.c7:focus:not(:focus-visible) >line,
.c7:focus:not(:focus-visible) >path,
.c7:focus:not(:focus-visible) >polygon,
.c7:focus:not(:focus-visible) >polyline,
.c7:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c7:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c6 {
    margin-top: 12px;
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      id="data"
    >
      <form
        class="c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <div
              class="c1"
            >
              <span
                class="c5"
              >
                1 item
              </span>
            </div>
          </div>
          <footer
            class="c6"
          >
            <button
              class="c7"
              type="submit"
            >
              Apply filters
            </button>
          </footer>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DataSummary theme separator margin and dataSummary margin 1`] = `
<DocumentFragment>
  .c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c4 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c8 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c14 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 0 auto;
  justify-content: center;
}

.c18 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  justify-content: center;
}

.c20 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  justify-content: center;
}

.c23 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  border: solid 2px #7D4CDB;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 24px;
  width: 24px;
  justify-content: center;
  border-radius: 4px;
}

.c26 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c28 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  border: solid 2px rgba(0, 0, 0, 0.15);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 24px;
  width: 24px;
  justify-content: center;
  border-radius: 4px;
}

.c5 {
  margin-top: 48px;
  margin-bottom: 48px;
  font-size: 18px;
  line-height: 24px;
}

.c27 {
  font-size: 18px;
  line-height: 24px;
  font-weight: bold;
}

.c7 {
  font-size: 18px;
  line-height: 24px;
}

.c6 {
  margin-left: 48px;
  margin-right: 48px;
  font-size: 18px;
  line-height: 24px;
}

.c9 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c9:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c9:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c9:focus >circle,
.c9:focus >ellipse,
.c9:focus >line,
.c9:focus >path,
.c9:focus >polygon,
.c9:focus >polyline,
.c9:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c9:focus ::-moz-focus-inner {
  border: 0;
}

.c9:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c9:focus:not(:focus-visible) >circle,
.c9:focus:not(:focus-visible) >ellipse,
.c9:focus:not(:focus-visible) >line,
.c9:focus:not(:focus-visible) >path,
.c9:focus:not(:focus-visible) >polygon,
.c9:focus:not(:focus-visible) >polyline,
.c9:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c9:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c2 {
  max-width: 100%;
}

.c24 {
  box-sizing: border-box;
  stroke-width: 4px;
  stroke: #7D4CDB;
  width: 24px;
  height: 24px;
}

.c19 {
  display: flex;
  flex-direction: row;
  user-select: none;
  -webkit-user-select: none;
  width: fit-content;
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 6px;
  padding-bottom: 6px;
  opacity: 0.5;
  cursor: default;
}

.c19:hover input:not([disabled])+div,
.c19:hover input:not([disabled])+span {
  border-color: #000000;
}

.c22 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
}

.c22:checked+span>span {
  left: calc( 48px - 24px );
  background: #7D4CDB;
}

.c21 {
  flex-shrink: 0;
}

.c12 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  text-align: inherit;
  width: 1px;
  overflow: hidden;
  text-align: start;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
}

.c13 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  text-align: inherit;
  text-align: start;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 6px;
  padding-bottom: 6px;
}

.c17 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  text-align: inherit;
  width: 1px;
  overflow: hidden;
  text-align: start;
}

.c25 {
  margin: 0;
  padding: 0;
  font-weight: inherit;
  text-align: inherit;
  text-align: start;
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 6px;
  padding-bottom: 6px;
}

.c10 {
  border-spacing: 0;
  border-collapse: collapse;
  width: inherit;
}

.c11 {
  position: relative;
  border-spacing: 0;
  border-collapse: separate;
}

.c16:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c16:focus >circle,
.c16:focus >ellipse,
.c16:focus >line,
.c16:focus >path,
.c16:focus >polygon,
.c16:focus >polyline,
.c16:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c16:focus ::-moz-focus-inner {
  border: 0;
}

.c16:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c16:focus:not(:focus-visible) >circle,
.c16:focus:not(:focus-visible) >ellipse,
.c16:focus:not(:focus-visible) >line,
.c16:focus:not(:focus-visible) >path,
.c16:focus:not(:focus-visible) >polygon,
.c16:focus:not(:focus-visible) >polyline,
.c16:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c16:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c15:focus {
  outline: 2px solid #6FFFB0;
}

.c15:focus:not(:focus-visible) {
  outline: none;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {
  .c8 {
    margin-top: 12px;
  }
}

@media only screen and (max-width: 768px) {
  .c23 {
    border: solid 2px #7D4CDB;
  }
}

@media only screen and (max-width: 768px) {
  .c28 {
    border: solid 2px rgba(0, 0, 0, 0.15);
  }
}

<div
    class="c0"
  >
    <div
      class="c1"
      id="data"
    >
      <form
        class="c2"
      >
        <div
          class="c3"
        >
          <div
            class="c4"
          >
            <div
              class="c1"
            >
              <span
                class="c5"
              >
                2 items
                <span
                  class="c6"
                >
                  |
                </span>
                <span
                  class="c7"
                >
                  1 selected
                </span>
              </span>
            </div>
          </div>
          <footer
            class="c8"
          >
            <button
              class="c9"
              type="submit"
            >
              Apply filters
            </button>
          </footer>
        </div>
      </form>
      <table
        class="c10 c11"
      >
        <thead
          class="StyledTable__StyledTableHeader-sc-1m3u5g-4 StyledDataTable__StyledDataTableHeader-sc-xrlyjm-5"
        >
          <tr
            class="StyledTable__StyledTableRow-sc-1m3u5g-2 "
          >
            <th
              class="c12 "
              scope="col"
            />
            <th
              class="c13 "
              id="grommet-data-table-header-name"
              scope="col"
            >
              <div
                class="c14"
              />
            </th>
            <th
              class="c13 "
              id="grommet-data-table-header-percent"
              scope="col"
            >
              <div
                class="c14"
              />
            </th>
          </tr>
        </thead>
        <tbody
          class="StyledTable__StyledTableBody-sc-1m3u5g-3 c15"
        >
          <tr
            class="StyledTable__StyledTableRow-sc-1m3u5g-2 c16"
          >
            <td
              aria-disabled="true"
              class="c17 "
            >
              <div
                class="c18"
                style="height: 0px;"
              >
                <label
                  class="c19"
                  disabled=""
                >
                  <div
                    class="c20 c21"
                    disabled=""
                  >
                    <input
                      aria-label="unselect Alan"
                      checked=""
                      class="c22"
                      disabled=""
                      type="checkbox"
                    />
                    <div
                      class="c23 "
                      disabled=""
                    >
                      <svg
                        class="c24"
                        disabled=""
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M6,11.3 L10.3,16 L18,6.2"
                          fill="none"
                        />
                      </svg>
                    </div>
                    <input
                      type="hidden"
                      value="true"
                    />
                  </div>
                </label>
              </div>
            </td>
            <th
              class="c25 "
              scope="row"
            >
              <div
                class="c26"
              >
                <span
                  class="c27"
                >
                  Alan
                </span>
              </div>
            </th>
            <td
              class="c25 "
            >
              <div
                class="c26"
              >
                <span
                  class="c7"
                >
                  20
                </span>
              </div>
            </td>
          </tr>
          <tr
            class="StyledTable__StyledTableRow-sc-1m3u5g-2 c16"
          >
            <td
              aria-disabled="true"
              class="c17 "
            >
              <div
                class="c18"
                style="height: 0px;"
              >
                <label
                  class="c19"
                  disabled=""
                >
                  <div
                    class="c20 c21"
                    disabled=""
                  >
                    <input
                      aria-label="select Bryan"
                      class="c22"
                      disabled=""
                      type="checkbox"
                    />
                    <div
                      class="c28 "
                      disabled=""
                    />
                  </div>
                </label>
              </div>
            </td>
            <th
              class="c25 "
              scope="row"
            >
              <div
                class="c26"
              >
                <span
                  class="c27"
                >
                  Bryan
                </span>
              </div>
            </th>
            <td
              class="c25 "
            >
              <div
                class="c26"
              >
                <span
                  class="c7"
                >
                  30
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
